src/allmydata/windows/fixups.py

```diff
@@ -240,3 +192,107 @@ def unmangle(s):
     sys.argv = argv[-len(sys.argv):]
     if sys.argv[0].endswith('.pyscript'):
         sys.argv[0] = sys.argv[0][:-9]
+
+
+def a_console(handle):
+    """
+    :return: ``True`` if ``handle`` refers to a console, ``False`` otherwise.
+    """
+    if handle == INVALID_HANDLE_VALUE:
+        return False
+    return (
+        # It's a character file (eg a printer or a console)
+        GetFileType(handle) == FILE_TYPE_CHAR and
+        # Checking the console mode doesn't fail (thus it's a console)
+        GetConsoleMode(handle, byref(DWORD())) != 0
+    )
+
+
+class UnicodeOutput(object):
+    """
+    ``UnicodeOutput`` is a file-like object that encodes unicode to UTF-8 and
+    writes it to another file or writes unicode natively to the Windows
+    console.
+    """
+    def __init__(self, hConsole, stream, fileno, name, _complain):
+        """
+        :param hConsole: ``None`` or a handle on the console to which to write
+            unicode.  Mutually exclusive with ``stream``.
+
+        :param stream: ``None`` or a file-like object to which to write bytes.
+
+        :param fileno: A result to hand back from method of the same name.
+
+        :param name: A human-friendly identifier for this output object.
+
+        :param _complain: A one-argument callable which accepts bytes to be
+            written when there's a problem.  Care should be taken to not make
+            this do a write on this object.
+        """
+        self._hConsole = hConsole
+        self._stream = stream
+        self._fileno = fileno
+        self.closed = False
+        self.softspace = False
+        self.mode = 'w'
+        self.encoding = 'utf-8'
+        self.name = name
+
+        self._complain = _complain
+
+        from allmydata.util.encodingutil import canonical_encoding
+        from allmydata.util import log
+        if hasattr(stream, 'encoding') and canonical_encoding(stream.encoding) != 'utf-8':
+            log.msg("%s: %r had encoding %r, but we're going to write UTF-8 to it" %
+                    (name, stream, stream.encoding), level=log.CURIOUS)
+        self.flush()
+
+    def isatty(self):
+        return False
+    def close(self):
+        # don't really close the handle, that would only cause problems
+        self.closed = True
+    def fileno(self):
+        return self._fileno
+    def flush(self):
+        if self._hConsole is None:
+            try:
+                self._stream.flush()
+            except Exception as e:
+                self._complain("%s.flush: %r from %r" % (self.name, e, self._stream))
+                raise
+
+    def write(self, text):
+        try:
+            if self._hConsole is None:
+                if isinstance(text, unicode):
```

Comment:
Shouldn't this be `if not isinstance(text, unicode)`?

Reply:
I don't think so.  If _hConsole is None then the data will be written to _stream instead.  _stream requires UTF-8-encoded bytes.

This was pre-existing code that got moved around a bit.  I probably wouldn't have structured things quite this way but I'd also rather not change them around too much in this PR either.